mediatek: add LED support for Xiaomi Redmi Router AX6000

HZFrodo opened the pull request:
V1:
1. The LED panel is on SPI1 bus, while SPI1 has a clock issue.
clk_disable_unused() function during the boot process will disable the sel_clk of SPI1 by mistake.
When driver trying to read registers of SPI1, CPU will hang.
So we enable the sel_clk explicitly to make SPI1 full functional.

2. LED panel has 2 physical RGB LEDs, which are controlled by a single wire proprietary protocal.
This device uses MOSI of SPI1 to output the single wire data to both LEDs.
With some kind of reverse-engineering method, I can set RGB colors for each of them finally.
And a new driver, leds-spi-single-wire, was written to light up the LED panel.

V2:
switch to multicolor led driver by @981213

ptpt52 commented:
mark

achterin commented:
nice work!

lgs2007m commented:
Good job! Thanks for you nice work!

lalasou commented:
very good

ziseon commented:
Wow, 2 great news (leds support and new partition layouts) for AX6000 at the same time!

lalasou reviewed (approved):
Very goodI hope it can be quickly merged into the main line.

iseeyou commented:
mark

981213 commented:
Is that a WS2812B (NeoPixel) LED? If that's true, it can do way more than just on and off.
Could someone try out the WS2812B protocol on this LED? There are many libraries available for Arduino. (e.g. https://fastled.io/)

981213 commented:
BTW there's a multicolor LED framework in Linux. If this is an RGB LED I'd prefer it implemented using that instead of registering multiple LEDs with hard-coded color value.

dhlun commented:
mark

HZFrodo commented:
> BTW there's a multicolor LED framework in Linux. If this is an RGB LED I'd prefer it implemented using that instead of registering multiple LEDs with hard-coded color value.

Yes, multicolor LED framework is available since kernel version 5.9, but the fact is that we lack of multicolor LED configuration framework in OpenWrt at the momment.
And what is OpenWrt going to do with 24bit colors? If we let our users to select colors at will, then a lot improvements of uci components are required to make multicolor LED work well. Currently, OpenWrt only need four LED states to indicate different status, see: https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh#L6
I am not very familiar to uci releated code, so this should be implemented by some other developer on a suppported target device.

981213 commented:
We want to avoid drivers which can't be submitted upstream if possible. Registering multiple LEDs that interfere with each other is clearly not acceptable upstream.

> I am not very familiar to uci releated code, so this should be implemented by some other developer on a suppported target device.

Adding a multi_intensity field to led scripts and default config isn't too hard. I could take over this part if you want :)

981213 commented:
I've written a driver for WS2812B (Tested on Raspberry Pi) and will be sending it upstream. If it's indeed the same protocol, you can also skip the driver part and wait for my patch :)

graysky2 commented:
@HZFrodo - If you try to merge your PR on top of PR#11115 it has conflicts.  Not sure if 11115 will get merged and if so before your PR.  Just FYI.

HZFrodo commented:
> I've written a driver for WS2812B (Tested on Raspberry Pi) and will be sending it upstream. If it's indeed the same protocol, you can also skip the driver part and wait for my patch :)

That is so nice! Would you create a new PR to make you driver available here? And I can test it on my AX6000. Then I will rebase my PR to utilise your driver

HZFrodo commented:
> @HZFrodo - If you try to merge your PR on top of PR#11115 it has conflicts. Not sure if 11115 will get merged and if so before your PR. Just FYI.

Since a new driver written by @981213 is coming, a rebase of my PR is required, so PR#11115 could merged first :)

981213 commented:
@HZFrodo I'll be sending it upstream and get it reviewed there before adding it to OpenWrt.
Meanwhile it would be great if you can test this early version on AX6000 to verify if it's the same protocol. Here's the source code:

[leds-ws2812b-spi.zip](https://github.com/openwrt/openwrt/files/10063628/leds-ws2812b-spi.zip)

and the current dt binding looks like this:

```dts
#include <dt-bindings/leds/common.h>
		spi@0 {
			status = "okay";
                        #address-cells = <1>;
                        #size-cells = <0>;
			ws2812: ws2812b@0 {
				#address-cells = <1>;
				#size-cells = <0>;
				compatible = "worldsemi,ws2812b-spi";
				reg = <0>;
				spi-max-frequency = <2850000>;
                                led@0 {
                                        reg = <0>;
                                        label = "rgb:led0";
					color-index = <LED_COLOR_ID_RED LED_COLOR_ID_GREEN LED_COLOR_ID_BLUE>;
					default-intensity = <10 50 90>;
                                        color = <LED_COLOR_ID_RGB>;
                                        function = LED_FUNCTION_STATUS;
					function-enumerator = <0>;
                                };
				led@1 {
                                        reg = <1>;
                                        color = <LED_COLOR_ID_RGB>;
                                        function = LED_FUNCTION_STATUS;
					function-enumerator = <1>;
                                };
			};
		};
```

HZFrodo commented:
> [leds-ws2812b-spi.zip](https://github.com/openwrt/openwrt/files/10063628/leds-ws2812b-spi.zip)
@981213 Your new driver works fine on AX6000, I can set different colors and different triggers.
Once this driver is added to OpenWrt, I will rebase my PR to utilise your driver, thanks.

graysky2 commented:
@HZFrodo @981213 - How can one try the driver?

graysky2 commented:
@HZFrodo - I would like to try this PR but have already merged [PR#11115](https://github.com/openwrt/openwrt/pull/11115).  Is it safe to just change `/mt7986a-xiaomi-redmi-router-ax6000.dts/mt7986a-xiaomi-redmi-router-ax6000.dtsi/` in this PR as-is?  I believe that is the target for the patch once you merge PR#11115.  Thanks.

ptpt52 commented:
@graysky2 
you may pick from https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c
and https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27

graysky2 commented:
@ptpt52 - thanks, works great!

sunnyday666 reviewed (commented):
This is perfect.

lalasou reviewed (commented):
ok

lalasou reviewed (commented):
ok

iseeyou reviewed (commented):
ok

981213 commented:
Please stop spamming this PR with empty review. We don't merge PR directly via Github. The "At least 6 approving reviewers" is only there to prevent accidental merge button clicking.

981213 commented:
Upstreaming the driver takes a bit longer than I thought...
@HZFrodo Would you mind taking my v4 patch here and update this PR?
https://patches.linaro.org/project/linux-leds/patch/20221212045558.69602-4-gch981213@gmail.com/

I'll then merge this PR and take care of replacing it with the final version when it got accepted upstream.
DT maintainer upstream suggests that default-intensity should not be a driver-specific property so I dropped it. The LED can only be white for now. I'll do something about the color part in the userspace after merging this PR.

graysky2 commented:
@981213 - Have you seen these from @ptpt52 :

> you may pick from https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c
and https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27

I picked those two commits and have blue, green, red, white, and yellow.

![85 ti](https://user-images.githubusercontent.com/534822/207300139-13d78b5e-1bd7-439b-b106-c50979c3e6a9.jpg)

graysky2 commented:
OK... so in the future, are you recommending that I not use the two commits from you but instead use this?

981213 commented:
> @981213 - Have you seen these from @ptpt52 :
> 
> > you may pick from [x-wrt@e614fad](https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c)
> > and [x-wrt@3c1b8dc](https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27)
> 
> I picked those two commits and have blue, green, red, white, and yellow.

It hard-coded some color values for this RGB LED chip and this approach will definitely be rejected on Linux upstream. I'd like to avoid carrying a local patch when there is an upstream-ready approach available (implementing its driver with the multicolor led interface).

HZFrodo commented:
> Upstreaming the driver takes a bit longer than I thought... @HZFrodo Would you mind taking my v4 patch here and update this PR? https://patches.linaro.org/project/linux-leds/patch/20221212045558.69602-4-gch981213@gmail.com/
> 
> I'll then merge this PR and take care of replacing it with the final version when it got accepted upstream. DT maintainer upstream suggests that default-intensity should not be a driver-specific property so I dropped it. The LED can only be white for now. I'll do something about the color part in the userspace after merging this PR.

No problem, I will take your pending patch and rework this PR :-)

HZFrodo committed:
kernel: kmod-leds-ws2812b: Add package

This adds the kmod-leds-ws2812b package pending upstream
https://patches.linaro.org/project/linux-leds/patch/20221212045558.69602-4-gch981213@gmail.com/
Some mods were applied to let this patch build success on kernel 5.15

Signed-off-by: Furong Xu <xfr@outlook.com>

HZFrodo committed:
mediatek: add LED support for Xiaomi Redmi Router AX6000

1. The LED panel is on SPI1 bus, while SPI1 has a clock issue.
    clk_disable_unused() function during the boot process will disable the sel_clk of SPI1 by mistake.
    When driver trying to read registers of SPI1, CPU will hang.
    So we enable the sel_clk explicitly to make SPI1 full functional.

2. LED panel has 2 WS2812B RGB LEDs.
    Since there is no RGB LED support in OpenWrt yet, only white color for now.

Signed-off-by: Furong Xu <xfr@outlook.com>

HZFrodo commented:
@981213 PR reworked.
Tested on my AX6000, all good.

graysky2 commented:
@HZFrodo - do I just need to pull this PR on top of HEAD to verify functionality?

HZFrodo commented:
> @HZFrodo - do I just need to pull this PR on top of HEAD to verify functionality?

Yes. But if you are using the previous version of my PR and do a sysupgrade with configuration retained, you have to edit the LED config section in /etc/config/system to use the new LED label, or LED will not light up.

981213 commented on target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts:
```diff
@@ -202,6 +214,35 @@
+&spi1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_led_pins>;
+	cs-gpios = <0>, <0>;
```
Does it really need a cs-gpios property?

HZFrodo replied:
Of course not, copied it from spi0 by mistake, I will delete it, thanks

981213 commented:
I turned the LED patch into a standalone package and split the commits. I've also dropped the heartbeat trigger because I'd prefer the status indicator to behave like other OpenWrt devices by default.
Merged. Thanks!

981213 commented:
BTW if anyone want to set a color for their router now, you can write the RGB value you want to the multi_intensity sysfs interface in /etc/rc.local

graysky2 commented:
> I've also dropped the heartbeat trigger because I'd prefer the status indicator to behave like other OpenWrt devices by default.

Oh no, I love the heartbeat settings.  Any chance to re-add?  Users are of course free to switch them on or not.

981213 commented:
> Oh no, I love the heartbeat settings. Any chance to re-add? Users are of course free to switch them on or not.

You can easily add that trigger in luci if you want it.
